Google.Api.Gax.Grpc/Rest/JsonStateTracker.cs

```diff
@@ -0,0 +1,240 @@
+                break;
+        }
+        return NextAction.BufferAndContinue;
+    }
+
+    private NextAction PushFromStringLiteralNoEscape(char c)
+    {
+        state = c == '"' ? State.WithinObject
+            : c == '\\' ? State.StringLiteralEscape
+            : State.StringLiteralNoEscape;
+        return NextAction.BufferAndContinue;
+    }
+
+    private NextAction PushFromStringLiteralAfterEscape(char c)
+    {
+        // We don't care what was escaped, so long as it was valid.
+        // (In the case of a hex escape, we don't even care about the
+        // hex digits following it.)
+        if (c == '"' || c == '\\' || c == '/' ||
+            c == 'b' || c == 'f' || c == 'n' ||
+            c == 'r' || c == 't' || c == 'u')
+        {
+            state = State.StringLiteralNoEscape;
+            return NextAction.BufferAndContinue;
+        }
+        state = State.Error;
+        return NextAction.SignalError;
+    }
+
+    private static bool IsJsonWhitespace(char c) =>
+        c == ' ' || c == '\t' || c == '\r' || c == '\n';
+
+    private enum State
+    {
+        /// <summary>
+        /// Before we first see [
+        /// We stay in this state, only accepting whitespace or [ until we see the first [
+        /// </summary>
+        BeforeTopArray,
+
+        /// <summary>
+        /// Between response objects.
+        /// In this state, we only accept:
+        /// - Whitespace (stay in this state)
+        /// - Comma (stay in this state)
+        /// - { (move to Normal state)
+        /// - ] (move to AfterTopArray state)
+        /// </summary>
+        JustInsideTopArray,
+
+        /// <summary>
+        /// Not in a string token, but somewhere within the top-level array.
+        /// </summary>
+        WithinObject,
+
+        /// <summary>
+        /// In a string token, but not directly after a backslash.
+        /// </summary>
+        StringLiteralNoEscape,
+
+        /// <summary>
+        /// In a string token, directly after a backslash.
+        /// (The only permitted characters to follow this are double-quote, backslash, slash, b, f, n, r, t or u.
+        /// Although u should then be followed by four hex digits, we don't enforce that.)
+        /// </summary>
+        StringLiteralEscape,
+
+        /// <summary>
+        /// We have detected an error. This is unrecoverable.
+        /// </summary>
+        Error,
+
+        /// <summary>
+        /// After the final ]
+        /// We can only accept whitespace after this.
+        /// </summary>
+        AfterTopArray
+    }
+
+    /// <summary>
+    /// The action that should be taken by the caller immediately after a call to <see cref="Push(char)"/>.
+    /// </summary>
+    internal enum NextAction
+    {
+        /// <summary>
+        /// We've received the closing ] at the end of the top level array.
+        /// There should be no further non-whitespace characters.
+        /// </summary>
+        SignalEndOfResponses,
```

Comment:
There's one little nit here re usage. If you signal end of response, then the caller is not "encouraged" to continue pushing characters, so that last piece of validation may not happen. But it's us using it, so maybe not a big deal.

The alternative is for this tokenizer to return here IgnoreAndContinue. We don't need to signal end of response as that's when the stream finishes, I think.

Reply:
I'm okay either way, to be honest. Let's see what @viacheslav-rostovtsev thinks.